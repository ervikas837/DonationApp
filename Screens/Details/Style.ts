import {StyleSheet, Dimensions} from 'react-native';
import {getFontFamily} from '../../src/assets/fonts/helper';

const screenHeight = Dimensions.get('window').height;

const style = StyleSheet.create({
  image: {
    width: '100%',
    height: screenHeight * 0.5,
  },
  textContainer: {
    padding: 16,
  },
  title: {
    fontFamily: getFontFamily('Inter', '700'),
    fontSize: 24,
  },
  location: {
    fontFamily: getFontFamily('Inter', '500'),
    fontSize: 18,
  },
  description: {
    fontFamily: getFontFamily('Inter', '500'),
    marginVertical: 16,
    fontSize: 18,
  },
  buttonTextStyle: {
    color: 'black',
    fontFamily: getFontFamily('Inter', '600'),
    fontSize: 18,
  },
});

export default style;
