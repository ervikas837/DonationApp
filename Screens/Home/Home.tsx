import {ActivityIndicator, FlatList, Text, View} from 'react-native';
import React, {FC, useEffect, useState} from 'react';
import Header from '../../components/Header/Header';
import ListItem from '../../components/ListItem/ListItem';
import style from './Style';
import CategoryItem from '../../components/CategoryItem/CategoryItem';
import useCategories from '../../hooks/useCategories';
import useItems from '../../hooks/useItems';
import {useSelector} from 'react-redux';
import {RootState} from '../../redux/store';
import {useNavigation} from '@react-navigation/native';
import {ScreenNavigationProp} from '../../types';

export type ListItemType = {
  title: string;
  location: string;
  image?: string;
  id: number;
  categoryId: number;
  description: string;
};

export type CategoriesType = {
  title: string;
  id: number;
};

const Home: FC = () => {
  const {
    categories,
    loading: categoriesLoading,
    error: categoriesErorr,
  } = useCategories();

  const {
    items: displayData,
    isLoading: itemsLoading,
    error: itemError,
  } = useItems();

  const user = useSelector((state: RootState) => state.user);
  const [data, setData] = useState(displayData);
  const [selectedCategoryId, setSelectedCategoryId] = useState(1);
  const navigation = useNavigation<ScreenNavigationProp<'Details'>>();

  useEffect(() => {
    setData(displayData);
  }, [displayData]);

  function onSelectCategory(id: number) {
    setSelectedCategoryId(id);
    if (id == 1) {
      setData(displayData);
    } else {
      const filteredData = displayData.filter(item => item.categoryId == id);
      setData(filteredData);
    }
  }

  function onSelectListItem(item: ListItemType) {
    navigation.navigate('Details', {item});
  }

  if (categoriesLoading || itemsLoading) {
    return (
      <ActivityIndicator
        testID="activity-indicator"
        size="large"
        color={'#0000ff'}
      />
    );
  }

  if (categoriesErorr || itemError) {
    return <Text>Error Loading State</Text>;
  }

  return (
    <View>
      <View style={style.listContainer}>
        <FlatList
          ListHeaderComponent={
            <>
              <Header name={user.firstName + ' ' + user.lastName} />
              <View style={style.categoriesContainer}>
                <FlatList
                  showsHorizontalScrollIndicator={false}
                  horizontal={true}
                  data={categories}
                  renderItem={({item}) => (
                    <CategoryItem
                      item={item}
                      onSelect={onSelectCategory}
                      isSelected={item.id == selectedCategoryId}
                    />
                  )}
                />
              </View>
            </>
          }
          showsVerticalScrollIndicator={false}
          keyExtractor={item => item.id.toString()}
          data={data}
          renderItem={({item}) => (
            <ListItem item={item} onSelect={onSelectListItem} />
          )}
        />
      </View>
    </View>
  );
};

export default Home;
