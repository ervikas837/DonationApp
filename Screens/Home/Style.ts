import {StyleSheet} from 'react-native';

const style = StyleSheet.create({
  categoriesContainer: {},
  listContainer: {
    alignItems: 'center',
  },
});

export default style;
