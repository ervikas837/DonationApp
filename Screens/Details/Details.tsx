import {View, Text, Image, Dimensions, ScrollView} from 'react-native';
import React, {FC, useEffect, useState} from 'react';
import style from './Style';
import {RouteProp, useIsFocused} from '@react-navigation/native';
import {ListItemType} from '../Home/Home';
import CustomButton from '../../components/Button/CustomButton';
import {Colors} from 'react-native/Libraries/NewAppScreen';
import {getFontFamily} from '../../src/assets/fonts/helper';

type DetailsScreenProps = {
  route: RouteProp<{params: {item: ListItemType}}>;
};

const Details: FC<DetailsScreenProps> = ({route}) => {
  const {item} = route.params;
  const [imageUri, setImageUri] = useState(item.image);

  useEffect(() => {
    setImageUri(item.image);
  }, [item]);

  function handleDonate() {}

  return (
    <View>
      <ScrollView testID="detail-scroll">
        <View>
          <Image
            key={item.id}
            style={style.image}
            source={{uri: imageUri}}
            resizeMode="cover"
            testID="item-image"
          />
        </View>
        <View style={style.textContainer}>
          <Text style={style.title}>{item.title}</Text>
          <Text style={style.location}>{item.location}</Text>
          <Text style={style.description}>{item.description}</Text>
        </View>
      </ScrollView>
      <CustomButton
        title="Donate"
        onPress={handleDonate}
        textStyle={style.buttonTextStyle}
        containerStyle={{width: '90%'}}
      />
    </View>
  );
};

export default Details;
