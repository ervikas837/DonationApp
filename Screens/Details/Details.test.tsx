import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import Details from './Details';
import {ListItemType} from '../Home/Home';
import {RouteProp} from '@react-navigation/native';
import {getFontFamily} from '../../src/assets/fonts/helper';

// title: string;
// location: string;
// image?: string;
// id: number;
// categoryId: number;
// description: string;

const mockItem: ListItemType = {
  title: 'This is orphnage',
  location: 'London',
  id: 1,
  categoryId: 1,
  image: '',
  description: 'this is just the test description',
};

const mockRoute: RouteProp<{params: {item: ListItemType}}, 'params'> = {
  key: 'details-key',
  name: 'params',
  params: {
    item: mockItem,
  },
  path: undefined, // Path is optional, so it can be undefined
};

describe('test the detail screen', () => {
  test('should display correct elements on the screen', () => {
    const {getByText} = render(<Details route={mockRoute} />);

    expect(getByText('This is orphnage')).toBeTruthy();
    expect(getByText('London')).not.toBeNull();
  });

  test('should update image uri when item prop is changes', () => {
    const newImageURL = 'https://picsum.photos/200/300/new_image.jpg';
    const {rerender, getByTestId} = render(<Details route={mockRoute} />);

    expect(getByTestId('item-image').props.source.uri).toBe(mockItem.image);

    const updatedItem = {...mockItem, image: newImageURL};
    rerender(<Details route={{...mockRoute, params: {item: updatedItem}}} />);

    expect(getByTestId('item-image').props.source.uri).toBe(newImageURL);
  });

  test('should call handle donate when donate button is pressed', () => {
    const {getByText} = render(<Details route={mockRoute} />);
    const donateButton = getByText('Donate');

    fireEvent.press(donateButton);

    expect(donateButton).toBeTruthy();
  });

  test('should match snapshot', () => {
    const tree = render(<Details route={mockRoute} />);
    expect(tree).toMatchSnapshot();
  });

  test('should contain scroll view for content', () => {
    const {getByTestId} = render(<Details route={mockRoute} />);
    const scrollView = getByTestId('detail-scroll');
    expect(scrollView).toBeTruthy();
  });

  test('should apply custom styles to the donate button', () => {
    const {getByText} = render(<Details route={mockRoute} />);
    const button = getByText('Donate');

    expect(button.props.style).toMatchObject([
      {
        color: 'black',
        fontFamily: getFontFamily('Inter', '600'),
        fontSize: 18,
      },
    ]);
  });
});
