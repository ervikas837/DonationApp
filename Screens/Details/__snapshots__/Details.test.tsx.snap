// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the detail screen should match snapshot 1`] = `
<View>
  <RCTScrollView
    testID="detail-scroll"
  >
    <View>
      <View>
        <Image
          resizeMode="cover"
          source={
            {
              "uri": "",
            }
          }
          style={
            {
              "height": 667,
              "width": "100%",
            }
          }
          testID="item-image"
        />
      </View>
      <View
        style={
          {
            "padding": 16,
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Inter-Bold",
              "fontSize": 24,
            }
          }
        >
          This is orphnage
        </Text>
        <Text
          style={
            {
              "fontFamily": "Inter-Medium",
              "fontSize": 18,
            }
          }
        >
          London
        </Text>
        <Text
          style={
            {
              "fontFamily": "Inter-Medium",
              "fontSize": 18,
              "marginVertical": 16,
            }
          }
        >
          this is just the test description
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#007BFF",
        "borderRadius": 25,
        "height": 48,
        "justifyContent": "center",
        "margin": 16,
        "opacity": 1,
        "width": "90%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "black",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 18,
          },
        ]
      }
    >
      Donate
    </Text>
  </View>
</View>
`;
